engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/typer/TypeMethodReference.scala

```diff
@@ -27,39 +25,64 @@ class TypeMethodReference(methodReference: MethodReference, currentResults: List
     ).toList
 
-  private def typeFromClazzDefinitions(clazzDefinitions: List[ClazzDefinition]): Either[String, TypingResult] =
-    clazzDefinitions match {
-      case Nil =>
-        Right(Unknown)
-      case _ =>
-        val isClass = clazzDefinitions.map(k => k.clazzName).exists(_.canBeSubclassOf(Typed[Class[_]]))
-        val display = clazzDefinitions.map(k => k.clazzName).map(_.display).mkString(", ")
-        clazzDefinitions.flatMap(_.methods.get(methodReference.getName)) match {
-          //Static method can be invoked - we cannot find them ATM
-          case Nil if isClass => Right(Unknown)
-          case Nil  => Left(s"Unknown method '${methodReference.getName}' in $display")
-          case methodInfoes => typeFromMethodInfoes(methodInfoes)
-        }
+  private def typeFromClazzDefinitions(clazzDefinitions: List[ClazzDefinition]): Either[String, TypingResult] = {
+    val validatedType = for {
+      nonEmptyClassDefinitions <- validateClassDefinitionsNonEmpty(clazzDefinitions).right
+      nonEmptyMethods <- validateMethodsNonEmpty(nonEmptyClassDefinitions).right
+      methodsWithMatchingArity <- validateMethodArity(nonEmptyMethods).right
+      returnTypesForMatchingParams <- validateMethodParameterTypes(methodsWithMatchingArity).right
+    } yield Typed(returnTypesForMatchingParams.toSet)
+
+    validatedType match {
+      case Left(None) => Right(Unknown) // we use Left(None) to indicate situation when we want to skip further validations because of lack of some knowledge
+      case Left(Some(message)) => Left(message)
+      case Right(returnType) => Right(returnType)
+    }
+  }
+
+  private def validateClassDefinitionsNonEmpty(clazzDefinitions: List[ClazzDefinition]): Either[Option[String], List[ClazzDefinition]] =
+    if (clazzDefinitions.isEmpty) Left(None) else Right(clazzDefinitions)
+
+  private def validateMethodsNonEmpty(clazzDefinitions: List[ClazzDefinition]): Either[Option[String], List[MethodInfo]] = {
+    def displayableType = clazzDefinitions.map(k => k.clazzName).map(_.display).mkString(", ")
+    def isClass = clazzDefinitions.map(k => k.clazzName).exists(_.canBeSubclassOf(Typed[Class[_]]))
+    clazzDefinitions.flatMap(_.methods.get(methodName)) match {
+      //Static method can be invoked - we cannot find them ATM
+      case Nil if isClass => Left(None)
+      case Nil => Left(Some(s"Unknown method '$methodName' in $displayableType"))
+      case methodInfoes => Right(methodInfoes)
     }
+  }
 
-  //TODO: we check only arity, but don't check if any of overloaded methods has matching signature
-  //this will lead to erros if we have different return types for different signatures!
-  private def typeFromMethodInfoes(methodInfoes: List[MethodInfo]): Either[String, TypingResult] =
-    methodInfoes.filter(_.parameters.size <= paramsCount) match {
+  private def validateMethodArity(methodInfoes: List[MethodInfo]): Either[Option[String], List[MethodInfo]] = {
```

Comment:
is this necessary now?

Reply:
If I will change `validateMethodParameterTypes` a bit, it won't be necessary. It can be a good idea because there is more details in error...